File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/java/src/main/java/com/example/graphalgorithms/Dijkstra.java
package com.example.graphalgorithms;

import java.util.*;

public class Dijkstra {
    public static class Edge {
        private final String to;
        private final int weight;

        public Edge(String to, int weight) {
            this.to = to;
            this.weight = weight;
        }

        public String to() { return to; }
        public int weight() { return weight; }
        public String toString() { return "(" + to + ", " + weight + ")"; }
    }

    private static class State {
        private final String vertex;
        private final int distance;

        State(String vertex, int distance) {
            this.vertex = vertex;
            this.distance = distance;
        }

        String vertex() { return vertex; }
        int distance() { return distance; }
    }

    public static Map<String, Integer> shortestPaths(Map<String, List<Edge>> graph, String start) {
        if (!graph.containsKey(start)) {
            throw new IllegalArgumentException("start vertex is not in the graph");
        }
        Map<String, Integer> distance = new HashMap<>();
        for (String vertex : graph.keySet()) {
            distance.put(vertex, Integer.MAX_VALUE / 4);
        }
        distance.put(start, 0);

        PriorityQueue<State> queue = new PriorityQueue<>(Comparator.comparingInt(State::distance));
        queue.add(new State(start, 0));
        while (!queue.isEmpty()) {
            State current = queue.poll();
            if (current.distance() > distance.get(current.vertex())) {
                continue;
            }
            for (Edge edge : graph.getOrDefault(current.vertex(), List.of())) {
                if (!graph.containsKey(edge.to())) {
                    throw new IllegalArgumentException("edge references an unknown vertex: " + edge.to());
                }
                if (edge.weight() < 0) {
                    throw new IllegalArgumentException("Dijkstra does not support negative edge weights");
                }
                int candidate = current.distance() + edge.weight();
                if (candidate < distance.getOrDefault(edge.to(), Integer.MAX_VALUE / 4)) {
                    distance.put(edge.to(), candidate);
                    queue.add(new State(edge.to(), candidate));
                }
            }
        }
        return distance;
    }

    public static void main(String[] args) {
        Map<String, List<Edge>> graph = Map.of(
            "A", List.of(new Edge("B", 4), new Edge("C", 2)),
            "B", List.of(new Edge("D", 1)),
            "C", List.of(new Edge("B", 1), new Edge("D", 5)),
            "D", List.of()
        );
        System.out.println("Dijkstra shortest paths from A");
        System.out.println(shortestPaths(graph, "A"));
    }
}

File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/python/graph_algorithms/dijkstra.py
from __future__ import annotations

import heapq
from math import inf


def dijkstra(graph: dict[str, list[tuple[str, int]]], start: str) -> dict[str, float]:
    if start not in graph:
        raise ValueError(f"start vertex {start!r} is not in the graph")
    distances = {vertex: inf for vertex in graph}
    distances[start] = 0
    heap = [(0, start)]

    while heap:
        current_distance, vertex = heapq.heappop(heap)
        if current_distance > distances[vertex]:
            continue
        for neighbor, weight in graph[vertex]:
            if neighbor not in graph:
                raise ValueError(f"edge references unknown vertex {neighbor!r}")
            if weight < 0:
                raise ValueError("Dijkstra does not support negative edge weights")
            distance = current_distance + weight
            if distance < distances.get(neighbor, inf):
                distances[neighbor] = distance
                heapq.heappush(heap, (distance, neighbor))
    return distances


if __name__ == "__main__":
    sample = {
        "A": [("B", 4), ("C", 2)],
        "B": [("C", 1), ("D", 5)],
        "C": [("D", 8), ("E", 10)],
        "D": [("E", 2)],
        "E": [],
    }
    print("Dijkstra shortest paths from A")
    print(dijkstra(sample, "A"))
